### PR TITLE
features/integration-of-the-programming-language-selector-widget-in-kuzya

### PR DESCRIPTION
**Pull request by OlehSeletskyi**

### CODE REVIEW

File: src/kuzya.cpp [hunk 1]
```diff
@@ -60,39 +66,7 @@ Kuzya::Kuzya(QWidget *parent)
...
-#else
-    toolBar->addSeparator();
-#endif
-    toolBar->addAction(actionUndo);
-    toolBar->addAction(actionRedo);
-    actionCut->setShortcuts(QKeySequence::Cut);
-    toolBar->addAction(actionCut);
-    actionCopy->setShortcuts(QKeySequence::Copy);
-    toolBar->addAction(actionCopy);
-    actionPaste->setShortcuts(QKeySequence::Paste);
```
Comment (alexchmykhalo) [inline src/kuzya.cpp:79]: You need to configure key sequences for actions.

File: src/kuzya.cpp [hunk 2]
```diff
@@ -183,14 +140,68 @@ Kuzya::Kuzya(QWidget *parent)
 
     RFileList = new QList<QString>();
 
+    myFile = nullptr;
```
Comment (alexchmykhalo) [inline src/kuzya.cpp:143]: Move to initializer list of Kuzya constructor.

File: src/kuzya.cpp [hunk 3]
```diff
@@ -183,14 +140,68 @@ Kuzya::Kuzya(QWidget *parent)
...
+
+    machine = new QStateMachine();
+    stateLanguageSelection = new QState(machine);
+    stateOfWritingCode = new QState(machine);
+
+    for(int i = 0; i < compiler->getSupportedLanguages().size(); i++)
```
Comment (alexchmykhalo) [inline src/kuzya.cpp]: You could rewrite it with range-based for.

File: src/kuzya.cpp [hunk 4]
```diff
@@ -405,17 +408,47 @@ void Kuzya::retranslate(void)
 *********openFile*****************************************************************************************************
 **/
 
+//void Kuzya::openFile(QString FileName)
```
Comment (alexchmykhalo) [inline src/kuzya.cpp]: Please, remove commented code.

File: src/kuzya.cpp [hunk 5]
```diff
@@ -666,7 +705,22 @@ void Kuzya::setUndoRedoEnabled()
     actionRedo->setEnabled(textEditor->isRedoAvailable());
 }
 
+void Kuzya::slotLanguageSelected(QString id)
+{
+    language = id;
+
+    myFile = new SourceFile();
+
+    qDebug() << "!!!temporaryFileName" << myFile->getFileName();
```
Comment (alexchmykhalo) [inline src/kuzya.cpp]: Please, remove debug output.

File: src/kuzya.cpp [hunk 6]
```diff
@@ -801,9 +855,65 @@ void Kuzya::refreshProfileSettings()
 /**
 *******************************************************************************************************
 **/
+//bool Kuzya::slotSave(void)
```
Comment (alexchmykhalo) [inline src/kuzya.cpp]: Please, remove commented code.

File: src/kuzya.cpp [hunk 7]
```diff
@@ -801,9 +855,65 @@ void Kuzya::refreshProfileSettings()
...
+//    file.close();
+//    statusBar()->showMessage(tr("Saved"), 2000);
+//    addFileNameToList(file.fileName());
+//    refreshProfileSettings();
+//    slotUpdateWindowName(false);
+//    if(settings->isLineMarginVisible) textEditor->setMarginWidth(3,QVariant(textEditor->lines()).toString());
+//    return true;
+//}
+
+void Kuzya::slotBeforeSave(void)
```
Comment (alexchmykhalo) [inline src/kuzya.cpp]: You could check for !myFile->isSaved() in slot save(). You do not need that slot Kuzya::slotBeforeSave(void)

File: src/kuzya.cpp [hunk 8]
```diff
@@ -801,9 +855,65 @@ void Kuzya::refreshProfileSettings()
...
+    {
+        fileName = QString::null;
+    }
+    slotSave();
+}
+
 bool Kuzya::slotSave(void)
 {
     QString newFileName;
+    qDebug() << "fileNameSlotSave" << fileName;
```
Comment (alexchmykhalo) [inline src/kuzya.cpp]: Please, remove unused debug output.

File: src/kuzya.cpp [hunk 9]
```diff
@@ -812,39 +922,49 @@ bool Kuzya::slotSave(void)
...
         }
-        delete msgBox;
-        return false ;
     }
-    QTextStream stream(&file);
-    stream << textEditor->text();
+
+    myFile->save(fileName);
+
+//    if(!myFile->isFileExists(fileName))
```
Comment (alexchmykhalo) [inline src/kuzya.cpp]: Please, remove commented code.

File: src/kuzya.cpp [hunk 10]
```diff
@@ -862,12 +982,26 @@ void Kuzya::slotSave_as(void)
...
+    }
 
     fileName = newFileName;
-    if (!slotSave()) fileName = oldFileName;
+
+    if (!slotSave())
+    {
+        fileName = oldFileName;
+    }
+//    if(!myFile->isSaved())
```
Comment (alexchmykhalo) [inline src/kuzya.cpp]: Please, remove commented code.

File: src/kuzya.cpp [hunk 11]
```diff
@@ -923,6 +1057,7 @@ void Kuzya::slotRun(void)
         slotCompile();
         compiler->waitForFinished(15000);
     }
+    qDebug() << "RUN";
```
Comment (alexchmykhalo) [inline src/kuzya.cpp]: Please, remove unused debug output.

File: src/kuzya.cpp [hunk 12]
```diff
@@ -1101,8 +1243,12 @@ bool Kuzya::slotSaveChangesNotifier(void)
         {
             return false;	// false - cancelbtn pressed
         }
-        if (msgBox->clickedButton() ==yesBtn)
+        if (msgBox->clickedButton() == yesBtn)
         {
+            if(!myFile->isSaved())
+            {
+                fileName = QString::null;
```
Comment (alexchmykhalo) [inline src/kuzya.cpp]: Use QString::clear().

File: src/optionsdialog.ui [hunk 13]
```diff
@@ -596,7 +596,7 @@ and Fonts</string>
    <item row="0" column="1">
```
Comment (alexchmykhalo) [inline src/optionsdialog.ui]: Please, ensure that you remove unused options.

File: src/sourcefile.cpp [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+
+    QString temporaryFileName = QUuid::createUuid().toString();
+
+    temporaryFileName.remove("{");
+    temporaryFileName.remove("}");
+    temporaryFileName.remove("-");
+
+    temporaryFileName = qApp->applicationDirPath()+ "/" + temporaryFileName + ".txt";
+
+    qDebug() << "SFtemporaryFileName" << temporaryFileName;
```
Comment (alexchmykhalo) [inline src/sourcefile.cpp]: Please, remove unused debug output.

File: src/sourcefile.cpp [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+        return true;
+    }
+    return false;
+}
+
+void SourceFile::save(QString fileName)
+{
+    if(!isSaved())
+    {
+        qDebug() << "SAVE_FileName" << mFile->fileName();
```
Comment (alexchmykhalo) [inline src/sourcefile.cpp]: Please, remove debug output.

**Review by alexchmykhalo — CHANGES_REQUESTED** [review]

File: src/kuzya.cpp (hunk 1 above)
Comment (OlehSeletskyi) [reply src/kuzya.cpp:79]: Set it up in the designer

File: src/kuzya.cpp (hunk 2 above)
Comment (OlehSeletskyi) [reply src/kuzya.cpp:143]: It is in the constructor of Kuzya. Am i wrong?

File: src/kuzya.cpp (hunk 3 above)
Comment (OlehSeletskyi) [reply src/kuzya.cpp]: Rewritten.

File: src/kuzya.cpp (hunk 4 above)
Comment (OlehSeletskyi) [reply src/kuzya.cpp]: Сompleted.

File: src/kuzya.cpp (hunk 5 above)
Comment (OlehSeletskyi) [reply src/kuzya.cpp]: Сompleted.

File: src/kuzya.cpp (hunk 6 above)
Comment (OlehSeletskyi) [reply src/kuzya.cpp]: Сompleted.

File: src/kuzya.cpp (hunk 7 above)
Comment (OlehSeletskyi) [reply src/kuzya.cpp]: Moved the check for !myFile->isSaved() in slotSave and removed slotBeforeSave.

File: src/kuzya.cpp (hunk 8 above)
Comment (OlehSeletskyi) [reply src/kuzya.cpp]: Сompleted.

File: src/kuzya.cpp (hunk 9 above)
Comment (OlehSeletskyi) [reply src/kuzya.cpp]: Сompleted.

File: src/kuzya.cpp (hunk 10 above)
Comment (OlehSeletskyi) [reply src/kuzya.cpp]: Сompleted.

File: src/kuzya.cpp (hunk 11 above)
Comment (OlehSeletskyi) [reply src/kuzya.cpp]: Сompleted.

File: src/kuzya.cpp (hunk 12 above)
Comment (OlehSeletskyi) [reply src/kuzya.cpp]: This is fileName = QString :: clear (); causes compilation error. Used fileName.clear ();

File: src/sourcefile.cpp (hunk 14 above)
Comment (OlehSeletskyi) [reply src/sourcefile.cpp]: Done.

File: src/sourcefile.cpp (hunk 15 above)
Comment (OlehSeletskyi) [reply src/sourcefile.cpp]: Done.

**Review by alexchmykhalo — APPROVED** [review]

**Pull request merged**